HW-6,7-MVVM패턴, AAC ViewModel, LiveData 적용

ironelder opened the pull request:
Presenter제거 및 ViewModel적용

ironelder committed:
HW-6-데이터 바인딩 적용을 위한 BaseViewModel 및 MainViewModel추가

HW-6-데이터 바인딩 적용을 위한 BaseViewModel 및 MainViewModel추가

ironelder committed:
HW-6-기존의 이미지 뷰 만 받는 BindingAdater에서 View를 받도록 수정

HW-6-기존의 이미지 뷰 만 받는 BindingAdater에서 View를 받도록 수정

ironelder committed:
HW-6-BaseFragment 코드 정리

HW-6-BaseFragment 코드 정리

ironelder committed:
HW-6-MainFragement에서 ViewModel사용하도록 수정

HW-6-MainFragement에서 ViewModel사용하도록 수정

ironelder committed:
HW-6-MainFragment에서 Presenter 제거 및 안쓰는 변수 제거

HW-6-MainFragment에서 Presenter 제거 및 안쓰는 변수 제거

ironelder committed:
HW-6-BaseFragment에서 Presenter제거 및 MainFragment에 적용

HW-6-BaseFragment에서 Presenter제거 및 MainFragment에 적용

ironelder committed:
HW-6-MainFragment에서 Destroy시에 Dispose클리어 할 수 있도록 수정

HW-6-MainFragment에서 Destroy시에 Dispose클리어 할 수 있도록 수정

ironelder committed:
HW-6-MainFragment의 layout에서 Databinding으로 사용하던 부분 수정

HW-6-MainFragment의 layout에서 Databinding으로 사용하던 부분 수정

ironelder committed:
HW-6-에러메시지 발생시 화면에 Toast를 띄우도록 적용

HW-6-에러메시지 발생시 화면에 Toast를 띄우도록 적용

ironelder committed:
HW-6-Presenter관련 클래스 및 파일과 사용하던 부분 의존성 제거 

HW-6-Presenter관련 클래스 및 파일과 사용하던 부분 의존성 제거

bean628 commented on ironelder/AndroidArchitecture/app/src/main/java/com/ironelder/androidarchitecture/view/mainview/MainFragment.kt:
```diff
@@ -67,15 +40,41 @@ class MainFragment :
+
+            })
+        }
     }
```
MainViewModel 생성할때  .apply를 이용해서 하시면  it?.  와 같이 널값을 체크하지 않아도 될것 같습니다~!

ironelder replied:
수정하였습니다~

3ef0da4

bean628 replied:
바인딩 변수에 있는 뷰모델 말고 처음에 바인드할때 생성하는곳에서 하는거 어떤가요??

ironelder replied:
뷰모델 선언시 apply를 통해서 초기 설정을 하도록 수정하였습니다~

81066f3

ironelder committed:
HW-6-mainViewModel에 apply적용 및 Null Safety추가

HW-6-mainViewModel에 apply적용 및 Null Safety추가

ironelder committed:
HW-6-ViewModel선언시 apply를 통해 데이터 초기 설정 사용하도록 설정

HW-6-ViewModel선언시 apply를 통해 데이터 초기 설정 사용하도록 설정

ironelder committed:
HW-7-AAC ViewModel적용을 위한 Base에 ViewModel상속

HW-7-AAC ViewModel적용을 위한 Base에 ViewModel상속

ironelder committed:
HW-7-AAC ViewModel을 위해 Observable변수를 LiveData로 변경

HW-7-AAC ViewModel을 위해 Observable변수를 LiveData로 변경

ironelder committed:
HW-7-Observable변수를 참조하고 있던 부분을 Live데이터의 형으로 수정

HW-7-Observable변수를 참조하고 있던 부분을 Live데이터의 형으로 수정

ironelder committed:
HW-7-View에서 ViewModel을 선언하고 사용하도록 개선

HW-7-View에서 ViewModel을 선언하고 사용하도록 개선

sjjeong commented on ironelder/AndroidArchitecture/app/src/main/java/com/ironelder/androidarchitecture/view/mainview/MainFragment.kt:
```diff
@@ -8,54 +8,50 @@ import android.view.MenuItem
-    }
-
     override fun doViewCreated(view: View, savedInstanceState: Bundle?) {
+        binding.lifecycleOwner = this@MainFragment
```
이 부분은 base로 빼면 좋습니다
또한 Fragment는 viewLifeCyclerOwner를 사용하는게 좋습니다.

BeokBeok replied:
[5 common mistakes when using Architecture Components](https://proandroiddev.com/5-common-mistakes-when-using-architecture-components-403e9899f4cb) 의 
`1. Leaking LiveData observers in Fragments` 참고하시면 좋을 것 같아요.

ironelder replied:
lifeCycleOwner를 Base로 옮기고 viewLifeCycleOwner를 사용하고, 선언 위치를 onActivityCreated로 옮겼습니다

b3c9236

sjjeong commented on ironelder/AndroidArchitecture/app/src/main/java/com/ironelder/androidarchitecture/view/mainview/MainFragment.kt:
```diff
@@ -8,54 +8,50 @@ import android.view.MenuItem
+        binding.mainViewModel =
+            ViewModelProviders.of(this@MainFragment)[MainViewModel::class.java].apply {
+                searchQuery.observe(this@MainFragment, Observer {
+                    searchWithAdapter(
```
searchQuery를 observe하고 여기서 다시 searchWithAdapter를 호출할 필요 없이
ViewModel 내부에서 처리해주시면 됩니다.

ironelder replied:
@sjjeong 
searchQuery를 제거하였습니다.
다만 ViewModel내부에서 처리하는 의미를 잘 모르겠습니다.

검색어 버튼 클릭하여 summit이 들어왔을때 searchWithAdapter을 호출하고 
ViewModel에서 searchQuery의 변수 값을 참조하여 사용하도록 수정하였는데
이게 맞는지 모르겠습니다

fba3558

sjjeong commented on ironelder/AndroidArchitecture/app/src/main/java/com/ironelder/androidarchitecture/view/mainview/MainFragment.kt:
```diff
@@ -8,54 +8,50 @@ import android.view.MenuItem
+                })
+                searchResultList.observe(this@MainFragment, Observer {
+
+                })
```
아무 동작하지 않는 코드입니다. 제거해주세요

ironelder replied:
제거하였습니다~
aa1fbdf

sjjeong commented on ironelder/AndroidArchitecture/app/src/main/java/com/ironelder/androidarchitecture/view/mainview/MainFragment.kt:
```diff
@@ -104,12 +97,19 @@ class MainFragment :
+            clearDisposable()
+        }
+    }
+
```
이 코드는 필요 없습니다.
내부적으로 ViewModel의 onCleared를 호출합니다.

ironelder replied:
해당 코드 제거하였습니다

aa1fbdf
936910f

ironelder committed:
HW-6,7-의미없는 ListObservable및 oncleared제거

HW-6,7-의미없는 ListObservable및 oncleared제거

ironelder committed:
HW-6,7-BaseViewModel에서 사용하지 않는 함수 제거

HW-6,7-BaseViewModel에서 사용하지 않는 함수 제거

ironelder committed:
HW-6,7-onActivityCreated추가 및 초기화 위치 변경, LifeCycleOwner를 this에서 변경

HW-6,7-onActivityCreated추가 및 초기화 위치 변경, LifeCycleOwner를 this에서 viewLifeCycleOwner로 변경

ironelder committed:
HW-6,7-searchQuery의 observe를 제거

HW-6,7-searchQuery의 observe를 제거

sjjeong commented on ironelder/AndroidArchitecture/app/src/main/java/com/ironelder/androidarchitecture/view/mainview/MainViewModel.kt:
```diff
@@ -1,62 +1,65 @@
-    override fun getSearchResultToRoom(
+    fun getSearchResultToRoom(
         type: String,
         searchResultDatabase: SearchResultDatabase?
```
SearchResultDatabase#getInstance에서 nullable로 오다보니 이와 관련된 코드들이 전부다 Nullable로 처리되고 있습니다.
getInstance를 nonnull로 변경하는게 좋을 것 같습니다.

ironelder replied:
SearchResultDatabase를 nonnull로 수정하였습니다

0b815d5

sjjeong commented on ironelder/AndroidArchitecture/app/src/main/java/com/ironelder/androidarchitecture/view/mainview/MainViewModel.kt:
```diff
@@ -1,62 +1,65 @@
         SearchDataRepositoryImpl.getRemoteSearchData(
             type,
-            query,
+            searchQuery.value,
```
getRemoteSearchData의 파라미터로 nullable를 받지 마시고 앞에서 null handling을 해주신 다음 nonnull로 전달해주세요

ironelder replied:
검색어 부분을 NonNull로 받도록 수정하였습니다.
0b7d8e0

sjjeong commented on ironelder/AndroidArchitecture/app/src/main/java/com/ironelder/androidarchitecture/view/mainview/MainFragment.kt:
```diff
@@ -70,12 +55,14 @@ class MainFragment :
+            getSearchResultToRoom(
+                mType ?: BLOG,
+                SearchResultDatabase.getInstance(
+                    context?.applicationContext ?: (activity as Context).applicationContext
```
activity도 nullable이기 때문에 as로 type casting하면 null 이어서 죽습니다. as?로 처리 하셔야 합니다.

ironelder replied:
해당 부분을 수정하였습니다
6440e40

@sjjeong 
수정하면서 activity를 as?로 하게 되면 뒤에 따라오는 applicationContext쪽에도 ? 를 붙여야 하는데 Room의 경우에는 applicationContext가 필요합니다.
앱 최초에 MainApplication에서 initialize함수를 선언하여 해당 부분을 이미 선언하고 있기는 합니다.
따라서 SearchResultDatabase 클래스에 
```
fun getInstance():SearchResultDatabase = INSTANCE!!
```
위와 같은 함수를 하나 더 추가해서 그냥 Instance만 받아와서 넘겨주도록 해야할지 잘 모르겠습니다

sjjeong replied:
이 부분은 koin 하면서 없어질 코드이기에 넘어가겠습니다

sjjeong reviewed (changes requested):
1. SearchResultDatabase에 SearchResultDao함수가 대문자로 시작하고 있습니다. 소문자로 시작하고 동사로 시작하게 해주세요
2. MainActivity에서 PagerAdapter가 미사용되고 있으니 제거 해주세요 
3. todo 처리 해주세요
4. reformat code 해주세요. 사용되지 않는 import제거와 코드 재정렬이 필요합니다.
5. Repository에서 LocalDataSource를 사용하지 않고 room database에 직접 접근하고 있습니다. LocalDataSource를 사용하게 변경해주세요

ironelder committed:
HW-6,7-코드 정리

HW-6,7-코드 정리

ironelder committed:
HW-6,7-SearchRemoteData에서 검색어 nullable 제거

HW-6,7-SearchRemoteData에서 검색어 nullable 제거

ironelder committed:
HW-6,7-MainActivity에서 PageAdapter제거

HW-6,7-MainActivity에서 PageAdapter제거

ironelder committed:
HW-6,7-Dao를 얻어오는 함수명이 대문자로 시작하는 부분 수정

HW-6,7-Dao를 얻어오는 함수명이 대문자로 시작하는 부분 수정

ironelder committed:
HW-6,7-context의 as를 nullable하도록 수정

HW-6,7-context의 as를 nullable하도록 수정

ironelder committed:
HW-6,7-SearchDataBase를 NonNull로 수정

HW-6,7-SearchDataBase를 nonnull로 수정

ironelder committed:
HW-6,7-Repository에서 Database에 직접 접근하지 않고 LocalRepository을 통하도록 수정

HW-6,7-Repository에서 Database에 직접 접근하지 않고 LocalRepository을 통하도록 수정

sjjeong reviewed (approved):
LGTM ✊ ✌️ 👋
